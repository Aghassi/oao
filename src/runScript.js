// @flow

import { removeAllListeners, addListener } from 'storyboard';
import parallelConsoleListener from 'storyboard-listener-console-parallel';
import { readAllSpecs } from './utils/readSpecs';
import { exec } from './utils/shell';
import { shortenName } from './utils/helpers';
import calcGraph from './utils/calcGraph';

type Options = {
  src: string,
  ignoreSrc?: string,
  tree?: boolean,
  parallel?: boolean,
  parallelLogs?: boolean,
  ignoreErrors?: boolean,
};

const run = async (
  script: string,
  { src, ignoreSrc, tree, parallel, parallelLogs, ignoreErrors }: Options
) => {
  if (parallel && parallelLogs) {
    removeAllListeners();
    addListener(parallelConsoleListener);
  }
  const allSpecs = await readAllSpecs(src, ignoreSrc, false);
  const pkgNames = tree ? calcGraph(allSpecs) : Object.keys(allSpecs);
  const allPromises = [];
  for (let i = 0; i < pkgNames.length; i++) {
    const pkgName = pkgNames[i];
    const { pkgPath, specs: prevSpecs } = allSpecs[pkgName];
    const storySrc = parallelLogs ? undefined : shortenName(pkgName, 20);
    if (prevSpecs.scripts && prevSpecs.scripts[script]) {
      let promise = exec(`yarn run ${script}`, {
        cwd: pkgPath,
        bareLogs: parallelLogs,
        storySrc,
      });
      if (ignoreErrors) promise = promise.catch(() => {});
      if (!parallel) {
        await promise;
      } else {
        allPromises.push(promise);
      }
    }
  }

  // If parallel logs are enabled, we have to manually exit.
  // We should also show the error again, since the parallel console
  // most probably swallowed it or only showed the final part.
  if (parallel && parallelLogs) {
    try {
      await Promise.all(allPromises);
    } catch (err) {
      if (err.stderr) {
        console.error(err.message); // eslint-disable-line
        console.error(err.stderr); // eslint-disable-line
        throw new Error(err.message);
      } else {
        throw err;
      }
    }
    process.exit(0);
  }
};

export default run;
