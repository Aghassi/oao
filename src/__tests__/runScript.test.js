/* eslint-env jest */
/* eslint-disable global-require */

import runScript from '../runScript';

jest.mock('../utils/shell');

describe('RUN-SCRIPT command', () => {
  it('executes the specified script on all sub-packages', async () => {
    const helpers = require('../utils/shell');
    await runScript('start', { src: 'test/fixtures/packages/*' });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });

  it('can do it following the dependency tree', async () => {
    const helpers = require('../utils/shell');
    await runScript('start', { src: 'test/fixtures/packages3/*', tree: true });
    expect(helpers.exec.mock.calls).toMatchSnapshot();
  });
});
