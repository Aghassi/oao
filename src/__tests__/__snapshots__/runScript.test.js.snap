// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RUN-SCRIPT command can do it following the dependency tree 1`] = `
Array [
  Array [
    "yarn run start",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages3/oao-b",
      "storySrc": "oao-b",
    },
  ],
  Array [
    "yarn run start",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages3/oao",
      "storySrc": "oao",
    },
  ],
  Array [
    "yarn run start",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages3/oao-c",
      "storySrc": "oao-c",
    },
  ],
  Array [
    "yarn run start",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages3/oao-d",
      "storySrc": "oao-d",
    },
  ],
  Array [
    "yarn run start",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages3/oao-priv",
      "storySrc": "oao-priv",
    },
  ],
]
`;

exports[`RUN-SCRIPT command executes the specified script on all sub-packages 1`] = `
Array [
  Array [
    "yarn run start",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages/oao-b",
      "storySrc": "oao-b",
    },
  ],
  Array [
    "yarn run start",
    Object {
      "bareLogs": undefined,
      "cwd": "test/fixtures/packages/oao-c",
      "storySrc": "oao-c",
    },
  ],
]
`;
